enhancements/machine-config/machine-config-irreconcilable-changes.md

```diff
@@ -0,0 +1,464 @@
+
+The [Workflow Description section](#workflow-description) give a deeper 
+explanation on the details of the summary.
+
+If future troubleshooting needs justifies the need of fine grain tracking the 
+enhancement, API and implementation will evolve in that direction.
+
+### Workflow Description
+
+Each time a MachineConfig is changed a new rendered MachineConfig is created
+for each pool associated to the changed MachineConfig.
+
+#### MachineConfigController early validation
+
+Before the freshly created rendered MachineConfig passes to the 
+MachineConfigDaemons the MCO performs an internal validation of it that can 
+can be divided into three phases:
+
+1. Parse the Ignition raw configuration.
+2. Ensure the Ignition configuration is valid.
+3. Ensure there are no changes to irreconcilable fields.
+
+The first two steps are self-explanatory and are not covered by this 
+enhancement, as the MCO will always perform them. The third one, the 
+validation of irreconcilable fields, is the main target of this enhancement.
+
+After the Ignition validation is done, the per-field irreconcilable validation 
+is performed or skipped based on the `irreconcilableFieldValidationOverrides` 
+list in the MachineConfiguration CR.
+
+```yaml
+apiVersion: operator.openshit.io/v1
+kind: MachineConfiguration
+metadata:
+    name: cluster
+spec:
+  # ...
+  irreconcilableFieldValidationOverrides:
+    - storageDisks       # Allow changes to Ignition's storage disks section
+    - storageFileSystems # Allow changes to Ignition's filesystem disks section
+# ...
+```
+
+The MCO, given, for example, the previous 
+`irreconcilableFieldValidationOverrides`, ensures no field but `disks` and 
+`filesystems` can be changed from the original install-time values and 
+a MachineConfig that was not possible till this enhancement, like the one 
+below can now progress:
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: MachineConfig
+metadata:
+  labels:
+    machineconfiguration.openshift.io/role: worker
+  name: 99-var-partition
+spec:
+  config:
+    ignition:
+      version: 3.5.0
+    storage:
+      disks:
+        - device: /dev/disk/by-path/pci-0000:3b:00.0-scsi-0:3:111:0
+          partitions:
+            - label: var
+              sizeMiB: 0
+              startMiB: 0
+          wipeTable: true
+      filesystems:
+        - device: /dev/disk/by-partlabel/var
+          format: xfs
+          path: /var
+```
+
+All validation fields not listed in `irreconcilableFieldValidationOverrides` 
+will continue to be validated against reconcilable changes as always.
+
+#### MachineConfigController irreconcilable changes summary report
+
+Before the MachineConfigController updates the nodes to point to the new 
+rendered configuration it creates the summary ConfigMap for the target pool, 
+that will be later referenced by the MCNs of nodes that contain 
+irreconcilable changes.
+
+The content of the ConfigMap for the previous example is as follows:
```

Comment:
Using a ConfigMap means that you can never perform any validation on this data and there is no guarantees to its structure.

Should this instead be a _new_ API that you use to store the data associated with these changes?

@JoelSpeed may have some additional thoughts on ConfigMap vs new CRD but I lean in the direction of a new CRD because:
- It enforces a guaranteed structure
- Validations can be written to ensure that only valid values are being provided
- We can make it have an empty spec and only populate the status, signaling to users this a "status type" meant only to report status of something
- Changes go through the standard API review process, which means it is less likely for breaking changes to happen to this structure

Reply:
In this particular case, this configmap is only used for history tracking purposes, and is not used in any existing workflow, nor is the user expected to ever interact with it short of debugging purposes. I think it would be fair to think of it as a persistent log entry.

Since we don't believe it is necessary to update this, I think a configmap for troubleshooting purposes should be sufficient.

Reply:
What should happen if someone manually modifies the data in the ConfigMap? What stops them from doing that?